src/dataframerow/utils.jl

```diff
@@ -145,50 +146,75 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
     return ngroups, rhashes, gslots, false
 end
 
-function row_group_slots(cols::Tuple{CategoricalVector},
+function row_group_slots(cols::NTuple{N,<:CategoricalVector},
                          hash::Val{false},
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool}
-    col = cols[1]
-    @assert groups === nothing || length(groups) == length(col)
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool} where N
+    # Computing neither hashes nor groups isn't very useful,
+    # and this method needs to allocate a groups vector anyway
+    @assert groups !== nothing && all(col -> length(col) == length(groups), cols)
 
     # If missings are to be skipped, they will all go to group 1,
     # which will be removed by group_rows
-    ngroups = length(levels(col)) + (eltype(col) >: Missing)
+    ngroupstup = map(cols) do c
+        length(levels(c)) + (!skipmissing && eltype(c) >: Missing)
+    end
+    ngroups = prod(ngroupstup) + skipmissing
+
+    # Fall back to hashing if there would be too many empty combinations.
+    # The first check ensures the computation of ngroups did not overflow.
+    # The rationale for the 2 threshold is that while the fallback method is always slower,
+    # it allocates a hash table of size length(groups) instead of the remap vector
+    # of size ngroups (i.e. the number of possible combinations) in this method:
+    # so it makes sense to allocate more memory for better performance,
+    # but it needs to remain reasonable compared with the size of the data frame.
+    if prod(Int128.(ngroupstup)) > typemax(Int) || ngroups > 2 * length(groups)
+        return invoke(row_group_slots,
+                      Tuple{Tuple{Vararg{AbstractVector}}, Val,
+                            Union{Vector{Int}, Nothing}, Bool},
+                      cols, hash, groups, skipmissing)
+    end
 
-    if groups !== nothing
+    seen = fill(false, ngroups)
+    # If skipmissing=true, missings will all go to group 1,
+    # which will be removed by group_rows
+    seen[1] = skipmissing
+    refmaps = map(cols) do col
         # When levels are in the same order as the index and there are no missing values,
-        # we could simply copy refs to groups, but the performance gain is negligible,
+        # we could simply use refs, but the performance gain is negligible,
         # so always sort groups in the order of levels
-        refmap = [0; CategoricalArrays.order(col.pool)]
-        seen = fill(false, length(refmap))
+        nlevels = length(levels(col))
+        refmap = Vector{Int}(undef, nlevels + 1)
+        refmap[1] = nlevels
+        refmap[2:end] .= CategoricalArrays.order(col.pool) .- 1
+        refmap
+    end
+    strides = (cumprod(collect(reverse(ngroupstup)))[end-1:-1:1]..., 1)::NTuple{N,Int}
```

Comment:
why do you need `collect` here? I guess it is not performance critical though.

Reply:
That's because `cumprod` doesn't accept tuples, unfortunately.